sdk/program/src/bignum.rs

```diff
@@ -0,0 +1,851 @@
+//! Solana BigNumber
+//!
+//! Rust-based BigNumber provides safe math for arbitrarily large numbers.
+//! Primarily focues on BPF program usage, non-BPF is also supported by
+//! backing from the OpenSSL implementation
+use std::fmt;
+
+#[repr(C)]
+#[derive(Debug)]
+/// BPF structure for sending/receiving BigNumber data to/from syscalls
+pub struct FfiBigNumber {
+    pub data_len: u64,     // Length of bytes of data file
+    pub is_negative: bool, // False (0) or True (1) if data is positive or negative
+    pub data: u64, // Pointer to the variable length big endian array represneting the vallue of BigNumber
+}
+
+#[repr(C)]
+#[derive(Eq, Ord, PartialOrd, AbiExample)]
+#[cfg(not(target_arch = "bpf"))]
+pub struct BigNumber {
+    value: openssl::bn::BigNum,
+}
+#[cfg(target_arch = "bpf")]
+pub struct BigNumber {
+    value: Vec<u8>,
+    is_negative: bool,
+}
+
+impl BigNumber {
+    /// Returns a BigNumber with initial value of 0
+    pub fn new() -> Self {
+        #[cfg(not(target_arch = "bpf"))]
+        {
+            let bn = openssl::bn::BigNum::new().unwrap();
+            Self { value: bn }
+        }
+        #[cfg(target_arch = "bpf")]
+        Self {
+            value: vec![0],
+            is_negative: false,
+        }
+    }
+
+    /// Returns the size, in bytes, of BigNum. Typically used in
+    /// assessing buffer size needed to fetch the bytes for serialization, etc.
+    pub fn size_in_bytes(&self) -> usize {
+        #[cfg(not(target_arch = "bpf"))]
+        {
+            self.value.num_bytes() as usize
+        }
+        #[cfg(target_arch = "bpf")]
+        {
+            self.value.len()
+        }
+    }
+
+    /// Returns a big-endian byte vector representation
+    /// of the absolute value of `self`
+    pub fn to_vec(&self) -> Vec<u8> {
```

Comment:
Be nice if there was an `as_ref` method for BPF only so folks wouldn't have to alloc/copy just to get the data values.  (or make `value` pub)

Reply:
Added